sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala

```diff
@@ -40,7 +40,8 @@ import org.apache.spark.util.{MutableURLClassLoader, Utils}
 /**
  * A class that holds all state shared across sessions in a given [[SQLContext]].
  */
-private[sql] class SharedState(val sparkContext: SparkContext) extends Logging {
+private[sql] class SharedState(val sparkContext: SparkContext, initConfig: Map[String, String])
```

Comment:
Hm, doesn't the problem apply to all the configuration references from `SparkContext` here?

Reply:
Yes, maybe all configuration in ShareState should  respect `initConfig` preferentially

Reply:
it seems we get 2 kinds of configs from `sparkContext`: `SparkConf` and hadoop `Configuration`. Maybe we can do
```
val conf = sparkContext.conf.clone()
val hadoopConf = sparkContext.hadoopConf.clone()
initConfig.foreach {
  // set conf and hadoopConf
}
```
Then we just use `conf` and `hadoopConf` in `SharedState`